### PR TITLE
PR#6279: add Set.map to stdlib

### PR DESCRIPTION
**Pull request by gasche**
[PR#6279](http://caml.inria.fr/mantis/view.php?id=6279)

### CODE REVIEW

File: stdlib/set.mli [hunk 1]
```diff
@@ -118,6 +118,17 @@ module type S =
        The elements of [s] are presented to [f] in increasing order
        with respect to the ordering over the type of the elements. *)
 
+    val map: (elt -> elt) -> t -> t
+    (** [map f s] creates a new set with elements [f a0],
```
Comment (dbuenzli) [inline stdlib/set.mli]: I would replace "creates a new set with elements" by "is a set with elements" (describe the value denoted by the expression).

File: stdlib/set.mli [hunk 2]
```diff
@@ -118,6 +118,17 @@ module type S =
...
+        [f a1]... [f aN], where [a0],[a1]..[aN] are the
+        values contained in [s].
+
+       The elements are passed to [f] in increasing order
+       with respect to the ordering over the type of the keys.
+
+       If [f] is the identity over the elements of [s], the returned
+       set is physically equal to [s]. *)
```
Comment (dbuenzli) [inline stdlib/set.mli]: Are you sure you want to provide this guarantee ?

Comment (gasche) [inline stdlib/set.mli]: This is done for the other functions for which the guarantee is provided. In the present case, it does constrain the possible implementation approaches (adding in a fold does not respect this requirement), which would be an argument against. But if I took the care of providing this guarantee in the implementation, I may as well let users rely on it.

Note that this is weaker that one may hope, for example it says nothing of sets were "almost all" elements are mapped to themselves, but some are changed.

File: stdlib/set.mli [hunk 3]
```diff
@@ -118,6 +118,16 @@ module type S =
        The elements of [s] are presented to [f] in increasing order
        with respect to the ordering over the type of the elements. *)
 
+    val map: (elt -> elt) -> t -> t
+    (** If the elements of [s] are [a0],[a1],..[aN], then the elements
+        of [map f s] are [f a0],[f a1]... [f aN].
```
Comment (dbuenzli) [inline stdlib/set.mli]: I personally prefer the documentation pattern "[map f s] is ..." it directly tells you how to read the occurence of such an expression in a program.

File: stdlib/moreLabels.mli [hunk 4]
```diff
@@ -158,6 +158,7 @@ module Set : sig
       val equal : t -> t -> bool
       val subset : t -> t -> bool
       val iter : f:(elt -> unit) -> t -> unit
+      val map : (elt -> elt) -> t -> t
```
Comment (rleonid) [inline stdlib/moreLabels.mli]: Shouldn't this be labeled ?

Comment (gasche) [inline stdlib/moreLabels.mli]: Indeed, thanks.

**Comment by alainfrisch** [comment]
- I would either make it explicit that the function must be the _physical_ identity on mapped elements, or test with `Ord.compare`rather than `==`.
- What about providing `filter_map` (with a function `'a -> 'a option') as well?

**Comment by gasche** [comment]
Using `Ord.compare` would be wrong if `map` intentionally update parts of the structure that are not taken into account by the comparison function (for example the values have a unique identifier, and `compare` uses them only, but the function updates another field). I'll update the specification.

**Comment by gasche** [comment]
I updated the branch.

**Comment by hcarty** [comment]
If we're going to add `filter_map` here it would be nice to add it to the rest of the appropriate stdlib modules as well.

**Comment by gasche** [comment]
I think I'm not going to work on `filter_map` as part of this pull-request.

**Pull request merged**